File: nff/commands/doctor.py
"""nff doctor — dependency and environment checks."""

import json
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

import click
import serial

import nff.tools.boards as boards_module
from nff import config
from nff.tools import arduino_lib, toolchain

_CLAUDE_DESKTOP_CONFIG = Path.home() / ".claude" / "claude_desktop_config.json"


@dataclass
class Check:
    passed: bool
    detail: str
    fix: Optional[str] = None
    optional: bool = False


def check_python() -> Check:
    vi = sys.version_info
    if vi.major >= 3 and vi.minor >= 10:
        return Check(passed=True, detail=f"Python {vi.major}.{vi.minor}.{vi.micro}")
    return Check(
        passed=False,
        detail=f"Python {vi.major}.{vi.minor} (need 3.10+)",
        fix="Install Python 3.10 or newer from python.org",
    )


def check_arduino_cli() -> Check:
    ver = toolchain.arduino_cli_version()
    if ver:
        return Check(passed=True, detail=ver)
    return Check(
        passed=False,
        detail="arduino-cli not found",
        fix="Run `nff install-deps` to install arduino-cli",
    )


def check_build_backend() -> Check:
    """The active build backend's tool: arduino-cli or PlatformIO."""
    backend = toolchain.active_backend()
    if backend == "platformio":
        from nff.tools.backends import platformio as pio
        ver = pio.platformio_version()
        if ver:
            return Check(passed=True, detail=f"platformio · {ver}")
        return Check(
            passed=False,
            detail="platformio not found",
            fix="Run `nff install-deps` to install PlatformIO",
        )
    arduino = check_arduino_cli()
    return Check(passed=arduino.passed, detail=f"arduino-cli · {arduino.detail}",
                 fix=arduino.fix, optional=arduino.optional)


def check_esptool() -> Check:
    ver = toolchain.esptool_version()
    if ver:
        return Check(passed=True, detail=ver)
    return Check(
        passed=False,
        detail="esptool not found",
        fix="Run `pip install esptool`",
        optional=True,
    )


def check_pyserial() -> Check:
    try:
        import serial as _s
        ver = getattr(_s, "__version__", "unknown")
        return Check(passed=True, detail=f"pyserial {ver}")
    except ImportError:
        return Check(passed=False, detail="pyserial not installed",
                     fix="Run `pip install pyserial`")


def check_config() -> Check:
    try:
        if not config.exists():
            return Check(passed=False, detail="~/.nff/config.json not found",
                         fix="Run `nff init` to create config")
        config.load()
        return Check(passed=True, detail=str(config.CONFIG_PATH))
    except config.ConfigError as exc:
        return Check(passed=False, detail=str(exc), fix="Run `nff init` to recreate config")


def check_device() -> Check:
    devices = boards_module.list_devices()
    if not devices:
        return Check(passed=False, detail="No known boards detected",
                     fix="Connect a supported board via USB")
    d = devices[0]
    try:
        conn = serial.Serial(d.port, timeout=1)
        conn.close()
        return Check(passed=True, detail=f"{d.board} on {d.port}")
    except serial.SerialException as exc:
        return Check(passed=False, detail=f"{d.board} on {d.port}: {exc}",
                     fix="Close other applications using the serial port")


def check_lib_sync() -> Check:
    """Informational: is the nff Arduino library synced, and is it stale vs a
    local nff-sdk-c checkout? Optional so it never flips the doctor exit code."""
    fields = arduino_lib.read_sync_meta()
    if not fields:
        return Check(
            passed=False,
            detail="nff Arduino library not synced",
            fix="Run `nff install-deps` (or `nff init`)",
            optional=True,
        )
    detail = (
        f"nff lib {fields.get('version', '?')} synced {fields.get('synced_at', '?')}"
    )
    warn = arduino_lib.local_sdk_newer_than_synced()
    if warn:
        return Check(passed=False, detail=f"{detail} — {warn}",
                     fix="Re-sync the nff library", optional=True)
    return Check(passed=True, detail=detail)


def check_claude_desktop() -> Check:
    path = _CLAUDE_DESKTOP_CONFIG
    if not path.exists():
        return Check(passed=False, detail=f"{path} not found",
                     fix="Run `nff init` to register nff as an MCP server", optional=True)
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return Check(passed=False, detail=f"{path} is not valid JSON", optional=True)
    if "nff" in data.get("mcpServers", {}):
        return Check(passed=True, detail="nff registered in claude_desktop_config.json")
    return Check(passed=False, detail="nff not found in mcpServers",
                 fix="Run `nff init` to register nff", optional=True)


@click.command()
def doctor():
    """Check that all nff dependencies are installed and configured."""
    checks = [
        ("Python", check_python()),
        ("Build backend", check_build_backend()),
        ("esptool", check_esptool()),
        ("pyserial", check_pyserial()),
        ("Config", check_config()),
    ]
    # The flattened nff Arduino library is an arduino-cli concept; the platformio
    # backend materialises the SDK per-project, so this check only applies there.
    if toolchain.active_backend() == "arduino":
        checks.append(("nff lib", check_lib_sync()))
    checks += [
        ("Device", check_device()),
        ("Claude Desktop", check_claude_desktop()),
    ]
    any_failed = False
    for name, ch in checks:
        icon = "✓" if ch.passed else ("?" if ch.optional else "✗")
        click.echo(f"  [{icon}] {name}: {ch.detail}")
        if not ch.passed and ch.fix:
            click.echo(f"        → {ch.fix}")
        if not ch.passed and not ch.optional:
            any_failed = True
    if any_failed:
        raise SystemExit(1)

File: nff-rs/nff/src/commands/doctor.rs
use crate::tools::{boards, config, toolchain};
use anyhow::Result;

struct Check {
    passed: bool,
    detail: String,
    fix: Option<String>,
    optional: bool,
}

impl Check {
    fn ok(detail: impl Into<String>) -> Self {
        Check { passed: true, detail: detail.into(), fix: None, optional: false }
    }
    fn fail(detail: impl Into<String>, fix: impl Into<String>) -> Self {
        Check { passed: false, detail: detail.into(), fix: Some(fix.into()), optional: false }
    }
    fn warn(detail: impl Into<String>, fix: impl Into<String>) -> Self {
        Check { passed: false, detail: detail.into(), fix: Some(fix.into()), optional: true }
    }
}

pub fn run() -> Result<()> {
    let checks = vec![
        check_arduino_cli(),
        check_esptool(),
        check_config(),
        check_device(),
        check_claude_desktop(),
        check_wokwi_cli(),
    ];

    let mut any_failed = false;

    for check in &checks {
        if check.passed {
            println!("  ✓ {}", check.detail);
        } else if check.optional {
            println!("  ⚠  {}", check.detail);
            if let Some(fix) = &check.fix {
                println!("    → {fix}");
            }
        } else {
            println!("  ✗ {}", check.detail);
            if let Some(fix) = &check.fix {
                println!("    → {fix}");
            }
            any_failed = true;
        }
    }

    if any_failed {
        std::process::exit(1);
    }
    Ok(())
}

fn check_arduino_cli() -> Check {
    match toolchain::arduino_cli_version() {
        Some(v) => Check::ok(format!(
            "{}  ({})",
            v,
            toolchain::find_arduino_cli().map(|p| p.display().to_string()).unwrap_or_default()
        )),
        None => Check::fail(
            "arduino-cli not found",
            "Install from https://arduino.github.io/arduino-cli",
        ),
    }
}

fn check_esptool() -> Check {
    match toolchain::esptool_version() {
        Some(v) => {
            let loc = toolchain::find_esptool()
                .map(|p| p.display().to_string())
                .unwrap_or_else(|| "python -m esptool".into());
            Check::ok(format!("{v}  ({loc})"))
        }
        None => Check::fail("esptool not found", "Run: pip install esptool"),
    }
}

fn check_config() -> Check {
    if !config::exists() {
        return Check::fail("Config not found", "Run: nff init");
    }
    match config::load() {
        Ok(_) => Check::ok(format!("Config found at {}", config::config_path().display())),
        Err(e) => Check::fail(
            format!("Config unreadable: {e}"),
            format!("Fix or delete {}", config::config_path().display()),
        ),
    }
}

fn check_device() -> Check {
    let devices = boards::list_devices();
    if devices.is_empty() {
        return Check::warn(
            "No recognised board detected",
            "Plug in a board and run `nff init`  (or use `nff flash --sim` / Wokwi tools without hardware)",
        );
    }
    let d = &devices[0];
    let sim = d.wokwi_chip.as_deref().unwrap_or("no Wokwi support");
    Check::ok(format!("Device detected: {} on {}  [sim: {}]", d.board, d.port, sim))
}

fn check_claude_desktop() -> Check {
    let cfg_path = dirs::home_dir()
        .unwrap_or_default()
        .join(".claude")
        .join("claude_desktop_config.json");

    if !cfg_path.exists() {
        return Check::fail("Claude Desktop config not found", "Run: nff init");
    }
    let raw = match std::fs::read_to_string(&cfg_path) {
        Ok(r) => r,
        Err(e) => return Check::fail(format!("Claude Desktop config unreadable: {e}"), ""),
    };
    let data: serde_json::Value = match serde_json::from_str(&raw) {
        Ok(v) => v,
        Err(e) => return Check::fail(format!("Claude Desktop config invalid JSON: {e}"), "Fix the file manually"),
    };
    if data["mcpServers"]["nff"].is_null() {
        return Check::fail("nff not registered in Claude Desktop config", "Run: nff init");
    }
    Check::ok(format!("Claude Desktop config OK  ({})", cfg_path.display()))
}

fn check_wokwi_cli() -> Check {
    match toolchain::wokwi_cli_version() {
        Some(v) => {
            let loc = toolchain::find_wokwi_cli()
                .map(|p| p.display().to_string())
                .unwrap_or_default();
            Check { passed: true, detail: format!("{v}  ({loc})"), fix: None, optional: true }
        }
        None => Check::warn(
            "wokwi-cli not found  (optional — required for --sim and nff wokwi)",
            "Install from https://github.com/wokwi/wokwi-cli",
        ),
    }
}
